### PR TITLE
[extractor/twitter] Fix unauthenticated extraction

### LINKED ISSUE #7473: [Twitter] Account is now required to download (Failed to parse JSON)
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I'm running yt-dlp version **2023.06.22** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

United States

### Provide a description that is worded well enough to be understood

As of today or yesterday, Twitter now requires an account to view tweets. As such, yt-dlp now breaks if username+password or cookies are not specified. I'm afraid there won't be any fix to this, and we'll have to make it error out when credentials are not specified.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://twitter.com/Twitter/status/1263145271946551300']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2023.06.30.182724 [af1fd12f6] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.19044-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg 2021-08-25-git-e41bd075dd-full_build-www.gyan.dev (setts), ffprobe 2021-08-25-git-e41bd075dd-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.18.0, brotli-1.0.9, certifi-2023.05.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.3
[debug] Proxy map: {}
[debug] Loaded 1854 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Available version: nightly@2023.06.30.182724, Current version: nightly@2023.06.30.182724
Current Build Hash: d11828df364bc4c5053fe96b816ebf614a81cbcf1491e12544d4853a2ed355dc
yt-dlp is up to date (nightly@2023.06.30.182724)
[twitter] Extracting URL: https://twitter.com/Twitter/status/1263145271946551300
[twitter] 1263145271946551300: Downloading guest token
[twitter] 1263145271946551300: Downloading GraphQL JSON
ERROR: [twitter] 1263145271946551300: 1263145271946551300: Failed to parse JSON (caused by JSONDecodeError("Expecting value in '': line 1 column 1 (char 0)")); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp\extractor\common.py", line 708, in extract
  File "yt_dlp\extractor\twitter.py", line 1148, in _real_extract
  File "yt_dlp\extractor\twitter.py", line 326, in _call_graphql_api
  File "yt_dlp\extractor\twitter.py", line 303, in _call_api
  File "yt_dlp\extractor\common.py", line 1062, in download_content
  File "yt_dlp\extractor\common.py", line 1032, in download_handle
  File "yt_dlp\extractor\common.py", line 1022, in parse
  File "yt_dlp\extractor\common.py", line 1009, in _parse_json
  File "yt_dlp\extractor\common.py", line 992, in __print_error

  File "yt_dlp\utils\_utils.py", line 624, in decode
  File "json\decoder.py", line 337, in decode
  File "json\decoder.py", line 355, in raw_decode
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "yt_dlp\extractor\common.py", line 1006, in _parse_json
  File "json\__init__.py", line 370, in loads
  File "yt_dlp\utils\_utils.py", line 632, in decode
json.decoder.JSONDecodeError: Expecting value in '': line 1 column 1 (char 0)
```

### PR DESCRIPTION
**Pull request by bashonly**
In response to Twitter authwalling all tweets. The owner claims this is temporary, so maybe we should wait a bit to see if he is telling the truth before merging this.

~~TODO:~~
   - ~~Update tests~~ done
   - ~~Include fix for Spaces?~~ separate PR
   - ~~Downgrade 'not all metadata' message severity: `report_warning` => `to_screen`~~ done

&nbsp;
Closes #7473

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🐦🔄🚀

Update Twitter extractors to use new APIs and simplify code. This improves the reliability and consistency of downloading videos and audio from Twitter and Twitter Spaces, and reduces code complexity and duplication.

> _We're sailing on the Twitter sea, with JSON and GraphQL_
> _We've left behind the legacy API, it was a bloody hell_
> _We've simplified the code, we've used the `traverse_obj` tool_
> _So heave away, me hearties, heave away and sing so cool_

### Walkthrough
*  Simplify the `_call_api` method of the `TwitterBaseIE` class and remove the guest token and retry logic ([link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103L288-R304))
*  Use the syndication JSON endpoint instead of the legacy API endpoint for extracting tweet information without authentication in the `_real_extract` method of the `TwitterIE` class ([link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103L1139-R1137))
*  Handle the possible 404 error from the syndication JSON endpoint and raise a login required error if the tweet is only available when logged in ([link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103L1139-R1137))
*  Import the `urllib.error` module to handle HTTP errors from the syndication JSON endpoint ([link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103R3))
*  Handle the case where the media ID is not present in the syndication JSON response and try to extract it from the video URL in the `extract_from_video_info` function ([link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103L1177-R1172))
*  Use the `traverse_obj` helper function to access nested fields in the JSON responses more safely and concisely in the `_real_extract` method of the `TwitterIE` class and the `extract_from_video_info` function ([link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103L1177-R1172), [link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103L1207-R1197), [link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103L1287-R1285))
*  Use the `mediaDetails` field instead of the `extended_entities` field in the syndication JSON response, as the latter may not be complete or present ([link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103L1287-R1285))
*  Remove the check for 'adult content' in the GraphQL errors, as it is no longer relevant for the syndication JSON method ([link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103L1082-L1083))
*  Add a check for login status in the `_real_extract` method of the `TwitterSpacesIE` class and raise a login required error if not logged in, as the GraphQL API for Twitter Spaces requires authentication ([link](https://github.com/yt-dlp/yt-dlp/pull/7476/files?diff=unified&w=0#diff-8b22c3c3c6185164ce8c3e82541f91c728397dec66456eef7e05f664d5729103R1472-R1473))

</details>

### CODE REVIEW

File: yt_dlp/extractor/twitter.py [hunk 1]
```diff
@@ -1136,19 +1115,22 @@ def _build_graphql_query(self, media_id):
...
-                'include_reply_count': 1,
-                'include_user_entities': 0,
-                'tweet_mode': 'extended',
-            }), 'retweeted_status', None)
+        if not self.is_logged_in:
+            try:
+                status = self._download_json(
+                    'https://cdn.syndication.twimg.com/tweet-result', twid, 'Downloading syndication JSON',
+                    headers={'User-Agent': 'Googlebot'}, query={'id': twid})
+                self.report_warning('Not all metadata is available without authentication')
```
Comment (pukkandan) [inline yt_dlp/extractor/twitter.py:1123]: to_screen is fine, but add `_LOGIN_HINT` too imo

**Review by pukkandan — APPROVED** [review]
> Include fix for Spaces?

It's fine to be in separate PR too

File: yt_dlp/extractor/twitter.py [hunk 2]
```diff
@@ -471,10 +453,10 @@ class TwitterIE(TwitterBaseIE):
             'timestamp': 1442188653,
             'upload_date': '20150913',
             'uploader_url': 'https://twitter.com/freethenipple',
-            'comment_count': int,
-            'repost_count': int,
+            'comment_count': None,  # only available with auth
+            'repost_count': None,  # same
```
Comment (pukkandan) [inline yt_dlp/extractor/twitter.py:457]:
```suggestion
            # comment_count, repost_count, view_count are only available with auth
```

All None check can be removed. This one comment is sufficient.

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by kapitaali** [comment]
Perfect fix, thank you

**Comment by gamer191** [comment]
Slightly off-topic, but I just noticed twitter videos can be downloaded by substituting the video id into `https://twitter.com/i/videos/tweet/1675748123468267521` (discovered by reverse engineering discord)

**Comment by bashonly** [comment]
> `https://twitter.com/i/videos/tweet/1675748123468267521`

I get a 403 ("bad guest token")

**Comment by gamer191** [comment]
I just realised that twitter's legacy API does work, if you change the authorisation token to (logged out example):
`authorization: Bearer AAAAAAAAAAAAAAAAAAAAAIK1zgAAAAAA2tUWuhGZ2JceoId5GwYWU5GspY4%3DUq7gzFoCZs1QfwGoVdvSac3IniczZEYXIcDyumCauIXpcAPorE`

**Comment by bashonly** [comment]
> `authorization: Bearer AAAAAAAAAAAAAAAAAAAAAIK1zgAAAAAA2tUWuhGZ2JceoId5GwYWU5GspY4%3DUq7gzFoCZs1QfwGoVdvSac3IniczZEYXIcDyumCauIXpcAPorE`

This does work with legacy API though. Hmm
It has a rate-limit of 180 (per ??), but the limit is per guest token, not per IP

**Comment by gamer191** [comment]
> This does work with the legacy API though. (& does _not_ work with graphql.)

Is it useful if it doesn't work with graphql?

**Comment by bashonly** [comment]
> Is it useful if it doesn't work with graphql?

GraphQL may be a lost cause for unauthenticated users, and the legacy API gives more metadata than the syndication endpoint, so yeah it is. It's just a question of which solution will be more stable going forward.

**Comment by pukkandan** [comment]
imo, let's merge this for now and wait for twitter to become stable

**Comment by bashonly** [comment]
> imo, let's merge this for now and wait for twitter to become stable

Yeah I think so too. Moving back to legacy API for tweet extraction basically means we have to revert to before 147e62fc584c3ea6fdb09bb7a47905df68553a22, and then add additional rate-limit handling. So it will be a whole other PR anyways

**Pull request merged**